cpp/src/copying/contiguous_split.cu

```diff
@@ -813,282 +907,302 @@ struct size_of_helper {
-  auto keys = cudf::detail::make_counting_transform_iterator(
-    0, [out_to_in_index] __device__(size_type i) { return out_to_in_index(i); });
-  auto values = thrust::make_transform_iterator(
-    d_dst_buf_info.begin(), [] __device__(dst_buf_info const& info) { return info.valid_count; });
-  thrust::reduce_by_key(rmm::exec_policy(stream),
-                        keys,
-                        keys + new_buf_count,
-                        values,
-                        thrust::make_discard_iterator(),
-                        dst_valid_count_output_iterator{_d_dst_buf_info});
-}
+  std::vector<uint8_t> h_buf_sizes_and_dst_info;
+  std::size_t* h_buf_sizes;
+  dst_buf_info* h_dst_buf_info;
+
+  rmm::device_buffer d_buf_sizes_and_dst_info;
+  std::size_t* d_buf_sizes;
+  dst_buf_info* d_dst_buf_info;
+};
+
+// Packed block of memory 3:
+// Pointers to source and destination buffers (and stack space on the
+// gpu for offset computation)
+struct packed_src_and_dst_pointers {
+  packed_src_and_dst_pointers(cudf::table_view const& input,
+                              std::size_t num_partitions,
+                              cudf::size_type num_src_bufs,
+                              rmm::cuda_stream_view stream,
+                              rmm::mr::device_memory_resource* mr)
+    : stream(stream),
+      src_bufs_size{cudf::util::round_up_safe(num_src_bufs * sizeof(uint8_t*), split_align)},
+      dst_bufs_size{cudf::util::round_up_safe(num_partitions * sizeof(uint8_t*), split_align)}
+  {
+    // host-side
+    h_src_and_dst_buffers = std::vector<uint8_t>(src_bufs_size + dst_bufs_size);
+    h_src_bufs            = reinterpret_cast<uint8_t const**>(h_src_and_dst_buffers.data());
+    h_dst_bufs = reinterpret_cast<uint8_t**>(h_src_and_dst_buffers.data() + src_bufs_size);
+
+    // device-side
+    d_src_and_dst_buffers = rmm::device_buffer(src_bufs_size + dst_bufs_size, stream, mr);
+    d_src_bufs            = reinterpret_cast<uint8_t const**>(d_src_and_dst_buffers.data());
+    d_dst_bufs            = reinterpret_cast<uint8_t**>(
+      reinterpret_cast<uint8_t*>(d_src_and_dst_buffers.data()) + src_bufs_size);
+
+    // setup src buffers
+    setup_src_buf_data(input.begin(), input.end(), h_src_bufs);
+  }
+
+  void copy_to_device()
+  {
+    CUDF_CUDA_TRY(cudaMemcpyAsync(d_src_and_dst_buffers.data(),
+                                  h_src_and_dst_buffers.data(),
+                                  src_bufs_size + dst_bufs_size,
+                                  cudaMemcpyDefault,
+                                  stream.value()));
+  }
+
+  rmm::cuda_stream_view const stream;
+  std::size_t const src_bufs_size;
+  std::size_t const dst_bufs_size;
+
+  std::vector<uint8_t> h_src_and_dst_buffers;
+  rmm::device_buffer d_src_and_dst_buffers;
+  uint8_t const** h_src_bufs;
+  uint8_t const** d_src_bufs;
+  uint8_t** h_dst_bufs;
+  uint8_t** d_dst_bufs;
+};
 
 };  // anonymous namespace
 
 namespace detail {
```

Comment:
Seems this is misplaced perhaps?
I think most of these functions should be in the anonymous namespace

Reply:
Though I normally see this pattern:

`cudf::foo` <- public API
`cudf::detail::foo` <- closely related impl

Reply:
That's right. The `detail` function would be something exposed in a `detail` header for another libcudf function to use internally. If the function is only used in this .cu file then just make it anonymous.

Reply:
So I have a constructor here that takes `stream` (chatting with @nvdbaranec on the side):

```
chunked_pack::chunked_pack(cudf::table_view const& input,
                           std::size_t user_buffer_size,
                           rmm::cuda_stream_view stream,
                           rmm::mr::device_memory_resource* mr);
```

this seems like I should remove stream from this non-detail API in the process.

Reply:
done

Reply:
I also made `mr` default to the per-device memory resource as other APIs do.